litedram/phy/ddr5/commands.py

```diff
@@ -0,0 +1,227 @@
+    ca : Array(2, Signal(14)), out
+        CA[13:0] values over 2 subsequent DRAM SDR clock cycles.
+    """
+
+    # String description of 1st and 2nd edge of each command, later parsed to
+    # construct the value. CS is assumed to be H for 1st edge and L for 2nd edge.
+    # TODO: Fix currently fake truth table to be consistent with the JEDEC specs for DDR5
+    # Here, in contrary to LPDDR4's table, first element of array is related to CS low,
+    # the second to CS high. Something will have to be changed, because other parts of code
+    # rely on order from LPDDR4, but there are commands that have only row related to low.
+    # At this moment I keep it as it is now.
+    TRUTH_TABLE = {
+        "ACTIVATE":      ["L L L R0 R1 R2 R3 BA0 BA1 BG0 BG1 BG2 CID0 CID1 CID2",
+                          "R4 R5 R6 R7 R8 R9 R10 R11 R12 R13 R14 R15 R16, CID3/R17"],
+        "READ":          ["H L H H H BL*=L BA0 BA1 BG0 BG1 BG2 CID0 CID1 CID2",
+                          "C2 C3 C4 C5 C6 C7 C8 C9 C10 V H V V CID3"],
+        "WRITE":         ["H L H H L BL*=L BA0 BA1 BG0 BG1 BG2 CID0 CID1 CID2",
+                          "V C3 C4 C5 C6 C7 C8 C9 C10 V H WR_partial=L V CID3"],
+        "PRECHARGE ALL": ["H H L H L CID3 V V V V L CID0 CID1 CID2"],
+        "REFRESH ALL":   ["H H L L H CID3 V V VorRIR VorH L CID0 CID1 CID2"],
+        "MPC":           ["H H H H L OP0 OP1 OP2 OP3 OP4 OP5 OP6 OP7 V"],
+        "MRR":           ["H L H L H MRA0 MRA1 MRA2 MRA3 MRA4 MRA5 MRA6 MRA7 V",
+                          "L L V V V V V V V V CW V V V"],
+        "MRW":           ["H L H L L MRA0 MRA1 MRA2 MRA3 MRA4 MRA5 MRA6 MRA7 V",
+                          "OP0 OP1 OP2 OP3 OP4 OP5 OP6 OP7 V V CW V V V"]
+    }
+
+    for cmd, (subcmd1, subcmd2) in TRUTH_TABLE.items():
+        assert len(subcmd1.split()) == 14, (cmd, subcmd1)
+        assert len(subcmd2.split()) == 14, (cmd, subcmd2)
+
+    def __init__(self, dfi_phase):
+        self.cs = Signal(2)
+        self.ca = Array([Signal(14), Signal(14)])  # CS high, CS low
+        self.dfi = dfi_phase
+
+    def set(self, cmd):
+        ops = []
+        for cyc, description in enumerate(self.TRUTH_TABLE[cmd]):
+            for bit, bit_desc in enumerate(description.split()):
+                ops.append(self.ca[cyc][bit].eq(self.parse_bit(bit_desc, is_mrw=cmd.startswith("MRW"))))
+        if cmd != "DESELECT":
+            ops.append(self.cs[0].eq(1))
+        return ops
+
+    def parse_bit(self, bit, is_mrw):
+        assert len(self.dfi.bank) >= 8, "At least 8 DFI bankbits needed for Mode Register address"
+        assert len(self.dfi.address) >= 17, "At least 17 DFI addressbits needed for row address"
+        mr_address = self.dfi.bank if is_mrw else self.dfi.address
+        rules = {
+            "H":       lambda: 1,  # high
+            "L":       lambda: 0,  # low
+            "V":       lambda: 0,  # defined logic
+            "X":       lambda: 0,  # don't care
+            "CID3/R17": None,
```

Comment:
I believe there cannot be `None` values.

Reply:
There are only temporary. I will replace them with lambdas, when I will know what they should return